lib/charms/mysql/v0/mysql.py

```diff
@@ -311,6 +330,242 @@ class RouterUser:
     router_id: str
 
 
+class MySQLCharmBase(CharmBase):
+    """Base class to encapsulate charm related functionality.
+
+    Meant as a means to share common charm related code between the MySQL VM and
+    K8s charms.
+    """
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.app_secrets, self.unit_secrets = None, None
+
+    def _on_get_password(self, event: ActionEvent) -> None:
+        """Action used to retrieve the system user's password."""
+        username = event.params.get("username") or ROOT_USERNAME
+
+        if username not in REQUIRED_USERNAMES:
```

Comment:
_since we are talking refactor_
suggestion: make `REQUIRED_USERNAME` a dict in constants mapping `usernames:keys`, e.g.:

```python
REQUIRED_USERS = {
  ROOT:ROOT_PASSWORD_KEY,
  ...
}
```

so this can be:
```suggestion
        if username not in REQUIRED_USERS.keys():
```

and follow up next comment:

Reply:
addressed in [c58177c](https://github.com/canonical/mysql-operator/pull/243/commits/c58177c6d12bbbb8a87248ff82ff3f140ce823fb)